src/realizations/catchment/Bmi_Py_Adapter.cpp

```diff
@@ -0,0 +1,247 @@
+                kw_arg = child.second.as_boolean();
+            }
+            else if (child.second.get_type() == geojson::PropertyType::Natural) {
+                kw_arg = child.second.as_natural_number();
+            }
+            else if (child.second.get_type() == geojson::PropertyType::Real) {
+                kw_arg = child.second.as_real_number();
+            }
+            // TODO: think about handling string explicitly, and handling other types; for now, document restrictions
+            else {
+                kw_arg = child.second.as_string();
+            }
+            kwargs = pybind11::dict(**kwargs, kw_arg);
+        }
+
+    }
+    else {
+        // If no __init__ args are present, just go with no-arg __init__
+        set_bmi_model(std::make_shared<pybind11::object>((*py_bmi_type_ref)()));
+    }
+    *********************************** * /
+
+    // Then perform the BMI initialization for the model
+    Initialize();
+}
+*/
+
+Bmi_Py_Adapter::Bmi_Py_Adapter(const string &type_name, string bmi_init_config, bool allow_exceed_end,
+                               bool has_fixed_time_step, const geojson::JSONProperty& other_input_vars,
+                               utils::StreamHandler output)
+    : Bmi_Py_Adapter(type_name, move(bmi_init_config), "", allow_exceed_end, has_fixed_time_step, other_input_vars,
+                     move(output)) {}
+
+Bmi_Py_Adapter::Bmi_Py_Adapter(const string &type_name, string bmi_init_config, string forcing_file_path,
+                               bool allow_exceed_end, bool has_fixed_time_step,
+                               const geojson::JSONProperty& other_input_vars, utils::StreamHandler output)
+                               : Bmi_Adapter<py::object>(type_name +" (BMI Py)", move(bmi_init_config),
+                                                         move(forcing_file_path), allow_exceed_end, has_fixed_time_step,
+                                                         other_input_vars, output),
+                                 np(py::module_::import("numpy")) /* like 'import numpy as np' */ { }
+
+string Bmi_Py_Adapter::GetComponentName() {
+    return py::str(bmi_model->attr("get_component_name")());
+}
+
+double Bmi_Py_Adapter::GetCurrentTime() {
+    // TODO: will need to verify the implicit casting for this works as expected
+    return py::float_(bmi_model->attr("get_current_time")());
+}
+
+double Bmi_Py_Adapter::GetEndTime() {
+    // TODO: will need to verify the implicit casting for this works as expected
+    return py::float_(bmi_model->attr("get_end_time")());
+}
+
+int Bmi_Py_Adapter::GetInputItemCount() {
+    return py::int_(bmi_model->attr("get_input_item_count")());
+}
+
+vector<string> Bmi_Py_Adapter::GetInputVarNames() {
+    vector<string> in_var_names(GetInputItemCount());
+    py::tuple in_var_names_tuple = bmi_model->attr("get_input_var_names")();
+    int i = 0;
+    for (auto && tuple_item : in_var_names_tuple) {
```

Comment:
this would almost be worth a helper function if you could parameterize the `py::` type.  Is there not a utility in pybind for converting to vector already?  Seems like that would be an easy and useful utility for that type of library.
```C++
template typename <T>
void tuple_to_vector(py::tuple tuple, std::vector<T>& vec)
{
int i = 0;
    for (auto && tuple_item : tuple) {
        T value = py::T(tuple_item);
        vec[i++] = value;
    }
}
```

Reply:
There seems to be auto conversion between `std::tuple` and `py::tuple`.  There isn't automatic converting to vector though, because tuples do not require all elements be of the same type.  We could add this kind of helper, but I think we'd need to consider it a little more carefully (e.g., should, and how would, the name reflect that it assumes all elements are of the same type in the tuple?).